### ISSUE TITLE
Best way to create accessible figures

### ISSUE THREAD

**Issue by bazzadp** [HTTPArchive/almanac.httparchive.org#1769]
Leading on from [a discussion from the Accessibility chapter](https://github.com/HTTPArchive/almanac.httparchive.org/pull/1752#pullrequestreview-550859976) I wanted to raise a separate issue fo this.

Currently we make a big effort to make our data accessible - and by that I don't just mean to assistive technology (AT) but to all readers.

To that end, we do the following:
- Every figure is given a "caption" which is a brief description or title of the figure. This is displayed to all users, but is also referenced from the figure through the `aria-labelledby` attribute.
- We also add an `alt` attribute which in most cases is the same as the caption, but has a slightly different purpose since we have the `aria-labelledby` attribute so from an accessibility point of view we don't need the `alt` attribute. So the `alt` attribute is used for the placeholder when an image fails to download and for SEO purposes (as they won't read `aria-labelledby` attributes). The `alt` attribute should also not have HTML tags that the caption may have (e.g. `<code>` tags or source links).
- Every figure is given a detailed "description" that is linked to the figure with `aria-describedby` attribute, which describes the figure in detail and, where the number of data points is small, we also include these. This is the one I want to discuss most on this, so will come back to this.
- In 2020 every figure should be directly linked to the data in Google Sheets and the SQL via the new context menu.

It's been raised that our Descriptions are possibly too descriptive, and would be better replaced with data tables instead of listing all the data points in a text description. Some chapters (e.g. the [2019 CDN chapter](https://almanac.httparchive.org/en/2019/cdn#fig-2)) included both, but in general we've discouraged this duplication and our Figures Guide goes as far as stating:

> Note: We do not insist on tables for all graphs (as some publications might for accessibility reasons), as the data sheets themselves are made available and linked from the chapters. Instead, accessibility is handled with descriptive alt attributes, ARIA labels and ARIA descriptions which are given in the chapters in the figure markup.

Personally I quite like what we have (but I'm massively biased as came up with a lot of this) and think the addition of the direct data links to Sheets and SQL further improves this. I can understand the request to include data tables instead of massively long descriptions but also feel the descriptions give us the opportunity to highlight the data to explain the impression of the data (e.g. in the 2020 CMS chapter we [have a description](https://almanac.httparchive.org/en/2020/cms#fig-3) stating "Bar chart showing WordPress dominating with 28.91% share in 2019, growing to 31.04% in 2020..."). But maybe we should not be as descriptive as we have been in the past and feel the need to list every data point?

What's people's thoughts on this? How can we improve this? I'd particularly love to hear any feedback from anyone with accessibility expertise (e.g. @alextait1 , @oluoluoxenfree , @ericwbailey @aardrian and anyone else with accessibility experience).

Also any thoughts on how to go about UX or Accessibility testing on the site?

**Comment by ericwbailey** [comment]
>We also add an alt attribute which in most cases is the same as the caption, but has a slightly different purpose since we have the aria-labelledby attribute so from an accessibility point of view we don't need the alt attribute.

This creates a double announce experience for some assistive technology, which isn't ideal. I'd recommend that the `figcaption` serve as the editorial description of what the content is, while the `alt` should be a literal description of the graph. Funnily enough, a coworker of mine [just wrote about this](https://thoughtbot.com/blog/alt-vs-figcaption).

I'm curious about the reports of "too descriptive." I don't think replacing them with a data table is a good solution, in that it puts the onus on an assistive tech user to traverse the table and mentally construct what the chart is illustrating. I do think, however, that linking to data tables would be great for any reader who wants to self serve.

For more complicated graphs that are serving an editorial purpose, I try to construct an alternate description that would be like if I tried to verbally describe to someone else the point the graph is trying to make. 

>also any thoughts on how to go about UX or Accessibility testing on the site?

I am a big proponent of CI/CD checks combined with a usability audit from a place like [Fable](https://www.makeitfable.com/). There's experiential stuff that an automated check won't catch, but a daily AT user can.

**Comment by bazzadp** [comment]
Thanks for chiming in @ericwbailey - great to hear your thoughts.

> > We also add an alt attribute which in most cases is the same as the caption, but has a slightly different purpose since we have the aria-labelledby attribute so from an accessibility point of view we don't need the alt attribute.
> 
> This creates a double announce experience for some assistive technology, which isn't ideal. I'd recommend that the `figcaption` serve as the editorial description of what the content is, while the `alt` should be a literal description of the graph. Funnily enough, a coworker of mine [just wrote about this](https://thoughtbot.com/blog/alt-vs-figcaption).

I'm curious about this? I would have thought an `aria-label` or `aria-labelledby` would automatically override the `alt`?

Or do you mean more the fact that we have an `alt`/`aria-labelledby` as well as a `figcaption`? If so, that's a good point and I don't know how to address this because we do want the descriptions to be available on-screen as well (not by default, but behind a button) so non-AT users can benefit from them too. So we've the image, the figcaption and the description - so there's going to be some repetition unfortunately, as don't want to leave the image with no label. I can't see how that can be avoided. Best we've come up with is to use `aria-labelledby` and `aria-describedby` on the image to point to the on-screen equivalents but presume that doesn't avoid the repetition.

> I'm curious about the reports of "too descriptive." I don't think replacing them with a data table is a good solution, in that it puts the onus on an assistive tech user to traverse the table and mentally construct what the chart is illustrating.

Was [suggested here](https://github.com/HTTPArchive/almanac.httparchive.org/pull/1752#pullrequestreview-550859976) particularly for small charts with few data points:

> One thought for the core team, maybe for next year: would be neat if instead of verbose text alternatives for bar charts if there were tables for AT users to understand the information.

And you were tagged further down that chat. I was concerned about that too, and with the link to the actual data think it's less necessary and personally prefer the descriptions. Plus means we can use them for non-data images and charts where there are too many data points to list.

> I do think, however, that linking to data tables would be great for any reader who wants to self serve.

Me too which is why we did it this year! Though in Google Sheets so not always the easiest to negotiate with AT I'm sure.

> For more complicated graphs that are serving an editorial purpose, I try to construct an alternate description that would be like if I tried to verbally describe to someone else the point the graph is trying to make.

Yeah we do try to do that. Particularly with those charts with lots of data points ([example](https://almanac.httparchive.org/en/2020/performance#fig-12)) but were trying to be nice and list the data points when there are only a few of them.

> > also any thoughts on how to go about UX or Accessibility testing on the site?
> 
> I am a big proponent of CI/CD checks combined with a usability audit from a place like [Fable](https://www.makeitfable.com/). There's experiential stuff that an automated check won't catch, but a daily AT user can.

We have [automated Lighthouse checks](https://calendar.perfplanet.com/2020/running-lighthouse-in-github-actions/) (uses AXE under the hood as I'm sure you are aware) and it's definitely helped us avoid introducing issues!

Not done any manual user testing or got an audit report yet though.

**Comment by svgeesus** [comment]
I agree with @ericwbailey that putting data points into accessible text can be problematic. Remember that sighted readers can skip ahead on text they don't want to read, while people using a screen reader can adjust the speed but access the content in a linear, sequential fashion so don't know if they are going to be read five numbers or 200.

Linking to data tables is great. People who want to visualize the data in a different way, analyze it further, compare it with other data, etc are well served by access to the original data. They can even raise issues if they feel the data has been mis-analyzed.

**Comment by bazzadp** [comment]
Closing this. Seems we have a pretty good implementation - and just need to be cautious not to overload the descriptions.
